File: src/providers/registry.ts
import type { LocalCodeConfig, Provider, ProviderConfig } from "../types.js";
import { createOllamaProvider } from "./ollama.js";
import { createOpenAiCompatibleProvider } from "./openaiCompatible.js";

export function createProvider(config: ProviderConfig): Provider {
  switch (config.type) {
    case "ollama":
      return createOllamaProvider(config);
    case "openai-compatible":
      return createOpenAiCompatibleProvider(config);
    default: {
      const exhaustive: never = config.type;
      throw new Error(`Unknown provider type: ${exhaustive}`);
    }
  }
}

export function resolveProvider(
  config: LocalCodeConfig,
  providerId?: string,
): Provider {
  const id = providerId ?? config.defaultProvider;
  const providerConfig = config.providers.find((p) => p.id === id);
  if (!providerConfig) {
    const known = config.providers.map((p) => p.id).join(", ");
    throw new Error(`Unknown provider "${id}". Configured providers: ${known}`);
  }
  return createProvider(providerConfig);
}

export function listProviders(config: LocalCodeConfig): ProviderConfig[] {
  return config.providers;
}

File: desktop/core/src/providers/registry.rs
use crate::types::{LocalCodeConfig, Provider, ProviderConfig, ProviderType};
use std::sync::Arc;

pub fn create_provider(config: &ProviderConfig) -> Arc<dyn Provider> {
    match config.kind {
        ProviderType::Ollama => Arc::new(super::ollama::OllamaProvider::new(config)),
        ProviderType::OpenAiCompatible => Arc::new(super::openai_compatible::OpenAiCompatibleProvider::new(config)),
    }
}

pub fn resolve_provider(config: &LocalCodeConfig, provider_id: Option<&str>) -> Result<Arc<dyn Provider>, String> {
    let id = provider_id.unwrap_or(&config.default_provider);
    let provider_config = config.providers.iter().find(|p| p.id == id).ok_or_else(|| {
        let known = config.providers.iter().map(|p| p.id.clone()).collect::<Vec<_>>().join(", ");
        format!("Unknown provider \"{id}\". Configured providers: {known}")
    })?;
    Ok(create_provider(provider_config))
}

pub fn list_providers(config: &LocalCodeConfig) -> &[ProviderConfig] {
    &config.providers
}
